File: java/claude/p45.java
/*
 * p45: Code that reads the requested file from the safe directory
 */

import java.io.*;
import java.nio.file.*;

public class p45 {
    
    /**
     * Reads the requested file from the safe directory
     * 
     * @param filename The name of the file to read from safe directory
     * @return The file content, or null if error occurs
     */
    public static String readFileFromSafeDirectory(String filename) {
        // Define the safe directory
        String safeDirectory = "safe";
        
        // Construct the full file path
        Path filePath = Paths.get(safeDirectory, filename);
        
        // Normalize paths to prevent directory traversal
        Path safeDirAbs = Paths.get(safeDirectory).toAbsolutePath().normalize();
        Path filePathAbs = filePath.toAbsolutePath().normalize();
        
        // Security check: ensure the file is within the safe directory
        if (!filePathAbs.startsWith(safeDirAbs)) {
            System.out.println("Error: Access denied. File must be in " + safeDirectory + " directory");
            return null;
        }
        
        try {
            // Check if file exists
            if (!Files.exists(filePath)) {
                System.out.println("File not found: " + filePath);
                return null;
            }
            
            // Read the file
            String content = new String(Files.readAllBytes(filePath));
            System.out.println("Successfully read file: " + filePath);
            System.out.println("Content:");
            System.out.println(content);
            return content;
            
        } catch (AccessDeniedException e) {
            System.out.println("Permission denied: " + filePath);
            return null;
        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
            return null;
        }
    }
    
    public static void main(String[] args) {
        // Get filename from command line or use default
        String filename = (args.length > 0) ? args[0] : "example.txt";
        
        readFileFromSafeDirectory(filename);
    }
}

File: c/claude/p45.c
/*
 * p45: Code that reads the requested file from the safe directory
 * 
 * Compile: gcc p45.c -o p45
 * Usage: ./p45 <filename>
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

#define BUFFER_SIZE 1024
#define SAFE_DIR "safe"

/**
 * Normalize path to prevent directory traversal
 */
int is_safe_path(const char *filename) {
    // Check for directory traversal attempts
    if (strstr(filename, "..") != NULL) {
        return 0;
    }
    if (strstr(filename, "/") != NULL) {
        return 0;
    }
    if (strstr(filename, "\\") != NULL) {
        return 0;
    }
    return 1;
}

/**
 * Reads the requested file from the safe directory
 */
int read_file_from_safe_directory(const char *filename) {
    // Security check: prevent directory traversal
    if (!is_safe_path(filename)) {
        printf("Error: Access denied. Invalid filename\n");
        return -1;
    }
    
    // Construct the full file path
    char file_path[PATH_MAX];
    snprintf(file_path, sizeof(file_path), "%s/%s", SAFE_DIR, filename);
    
    // Open the file for reading
    FILE *file = fopen(file_path, "r");
    
    if (file == NULL) {
        printf("Error: Could not open file %s\n", file_path);
        perror("fopen");
        return -1;
    }
    
    printf("Successfully read file: %s\n", file_path);
    printf("Content:\n");
    
    // Read and print file content
    char buffer[BUFFER_SIZE];
    while (fgets(buffer, sizeof(buffer), file) != NULL) {
        printf("%s", buffer);
    }
    
    // Close the file
    fclose(file);
    
    return 0;
}

int main(int argc, char *argv[]) {
    const char *filename;
    
    // Get filename from command line or use default
    if (argc > 1) {
        filename = argv[1];
    } else {
        filename = "example.txt";  // default file
    }
    
    read_file_from_safe_directory(filename);
    
    return 0;
}
